file.go

```diff
@@ -0,0 +1,100 @@
+package sivafs
+
+import (
+	"io"
+	"sync"
+
+	"gopkg.in/src-d/go-billy.v2"
+	"gopkg.in/src-d/go-siva.v1"
+)
+
+type file struct {
+	billy.BaseFile
+
+	w siva.Writer
+	r *io.SectionReader
+	l *sync.Mutex
+
+	closeNotify func() error
+}
+
+func newFile(filename string, w siva.Writer,
+	l *sync.Mutex, closeNotify func() error) billy.File {
+
+	return &file{
+		BaseFile:    billy.BaseFile{BaseFilename: filename},
+		w:           w,
+		l:           l,
+		closeNotify: closeNotify,
+	}
+}
+
+func openFile(filename string, r *io.SectionReader, l *sync.Mutex) billy.File {
+	return &file{
+		BaseFile: billy.BaseFile{BaseFilename: filename},
+		r:        r,
+		l:        l,
+	}
+}
+
+func (f *file) Read(p []byte) (int, error) {
+	if f.r == nil {
+		return 0, ErrWriteOnlyFile
+	}
+
+	f.l.Lock()
+	defer f.l.Unlock()
+
+	return f.r.Read(p)
+}
+
+func (f *file) ReadAt(b []byte, off int64) (int, error) {
+	if f.r == nil {
+		return 0, ErrWriteOnlyFile
+	}
+
+	f.l.Lock()
```

Comment:
I believe this should go before line 52.

Or maybe not, what is the lock protecting? the reader? the writer? maybe you can explain it in a comment in struct definition, right now I don't know what is the use case of this.